File: server/src/messageStore.js
const fs = require('fs');
const path = require('path');
const config = require('./config');

const LOG_FILE = path.join(path.dirname(config.sessionFile), 'messages.jsonl');
const MAX_BUFFER = 1000;

let buffer = [];

function loadRecent() {
  try {
    if (fs.existsSync(LOG_FILE)) {
      const lines = fs.readFileSync(LOG_FILE, 'utf8').trim().split('\n').filter(Boolean);
      buffer = lines.slice(-MAX_BUFFER).map((line) => JSON.parse(line));
      console.log(`Loaded ${buffer.length} message(s) from history.`);
    }
  } catch (err) {
    console.warn('Could not load message history:', err.message);
  }
}

function add(message) {
  buffer.push(message);
  if (buffer.length > MAX_BUFFER) buffer.shift();
  try {
    fs.mkdirSync(path.dirname(LOG_FILE), { recursive: true });
    fs.appendFileSync(LOG_FILE, JSON.stringify(message) + '\n');
  } catch (err) {
    console.warn('Could not persist message:', err.message);
  }
}

function update(message) {
  const idx = buffer.findIndex((m) => m.id === message.id);
  if (idx >= 0) {
    buffer[idx] = message;
  } else {
    buffer.push(message);
  }
}

function getRecent(limit = 200) {
  return buffer.slice(-limit).reverse(); // newest first
}

module.exports = { loadRecent, add, update, getRecent };

File: python-app/message_store.py
import json

import config

MAX_BUFFER = 1000
_buffer = []


def _log_file():
    return config.DATA_DIR / 'messages.jsonl'


def load_recent():
    global _buffer
    log_file = _log_file()
    if not log_file.exists():
        return
    try:
        lines = log_file.read_text(encoding='utf-8').strip().splitlines()
        _buffer = [json.loads(line) for line in lines[-MAX_BUFFER:] if line]
        print(f'Loaded {len(_buffer)} message(s) from history.')
    except Exception as err:
        print(f'Could not load message history: {err}')


def add(message):
    _buffer.append(message)
    if len(_buffer) > MAX_BUFFER:
        _buffer.pop(0)
    try:
        config.DATA_DIR.mkdir(parents=True, exist_ok=True)
        with open(_log_file(), 'a', encoding='utf-8') as f:
            f.write(json.dumps(message) + '\n')
    except Exception as err:
        print(f'Could not persist message: {err}')


def update(message):
    for i, existing in enumerate(_buffer):
        if existing['id'] == message['id']:
            _buffer[i] = message
            return
    _buffer.append(message)


def get_recent(limit=200):
    return list(reversed(_buffer[-limit:]))
